### PR TITLE
New machine drop

### PR DESCRIPTION
**Pull request by huwshimi**
Create new machines when units are dropped on machine view column headers.

### CODE REVIEW

**Comment by huwshimi** [comment]
QA:
- add a service
- open the machine view
- start dragging
- the machine column header should show a drop target
- drop the unit on the header
- the new machine form should appear
- click Place
- add another unit
- start dragging
- the container column header should now also show a drop target
- drop the unit on the container header
- the create container form should appear
- select a container type and click Place

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4003/
Test PASSed.

File: jujugui/static/gui/src/app/components/machine-view/test-machine-view.js [hunk 1]
```diff
@@ -21,12 +21,29 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
...
 
 describe('MachineView', function() {
+  var machines;
 
   beforeAll(function(done) {
     // By loading this file it adds the component to the juju components.
     YUI().use('machine-view', function() { done(); });
   });
 
+  beforeEach(function() {
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/machine-view/test-machine-view.js]: \o/

Glad to see all that duplicated code cleaned up.

File: jujugui/static/gui/src/app/components/machine-view/test-machine-view.js [hunk 2]
```diff
@@ -320,12 +329,13 @@ describe('MachineView', function() {
         addGhostAndEcsUnits={sinon.stub()}
         autoPlaceUnits={autoPlaceUnits}
         createMachine={createMachine}
+        destroyMachines={sinon.stub()}
         environmentName="My Env"
-        units={units}
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/machine-view/test-machine-view.js]: Thanks for alphabetizing these; as the list of props gets more unwieldy it'll be nice to have that in place.

**Comment by kadams54** [comment]
QA is OK. :+1: for the code.

File: jujugui/static/gui/src/app/components/machine-view/machine-view.js [hunk 3]
```diff
@@ -106,8 +106,18 @@ YUI.add('machine-view', function() {
...
     */
-    _dropUnit: function(unit, machine) {
-      this.props.placeUnit(unit, machine);
+    _dropUnit: function(unit, machine, newType) {
+      if (machine) {
+        this.props.placeUnit(unit, machine);
+      } else {
+        var state = {};
+        if (newType === 'machine') {
+          state.showAddMachine = {unit: unit};
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/machine-view/machine-view.js]: `showAddMachine` and `showAddContainer` are Boolean when defined but here you're passing an object. It's typically considered bad practice to change types. Can this be reworked to not change the type?

**Comment by hatched** [comment]
:+1: :ok: 
This is awesome, thanks for sticking through on this one. Would be nice if you could tackle that Type issue before landing.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4019/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4020/
Test PASSed.

**Comment by huwshimi** [comment]
Thanks for the reviews @kadams54 and @hatched :shipit:!

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**